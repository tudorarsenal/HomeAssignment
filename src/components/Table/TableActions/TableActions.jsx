import { useEffect, useRef, useCallback } from 'react';
import PropTypes from 'prop-types';

import { ReactComponent as DonwloadIcon } from '../icons/downloadIcon.svg';
import {
  SELECTED,
  NO_ROWS_SELECTED,
  DOWNLOAD_LABEL,
  SELECT_ALL,
} from '../utils/textsEn';

import './TableActions.scss';

const TableActions = ({
  isSelectable,
  isDownloadable,
  selectedRowsCount,
  onSelectAll,
  onDownload,
  areAllRowsSelected,
  downloadRules,
  areDownloadRulesMeet,
}) => {
  const selectAllRef = useRef();

  // true if some of the rows are selected
  // not all or not 0 selected
  const isIndeterminateValue = useCallback(
    () => isSelectable && !areAllRowsSelected === (selectedRowsCount !== 0),
    [isSelectable, areAllRowsSelected, selectedRowsCount]
  );

  useEffect(() => {
    if (!selectAllRef?.current) return;

    if (isIndeterminateValue()) {
      selectAllRef.current.indeterminate = true;
      return;
    } else if (isSelectable) {
      selectAllRef.current.indeterminate = false;
    }
  }, [isIndeterminateValue, isSelectable]);

  return (
    <div className="table-actions">
      {isSelectable && (
        <div className="table-actions__select">
          <label htmlFor="select-all" className="visually-hidden">
            {SELECT_ALL}
          </label>
          <input
            type="checkbox"
            id="select-all"
            name="select-all"
            checked={areAllRowsSelected}
            onChange={onSelectAll}
            ref={selectAllRef}
            className={`${
              isIndeterminateValue()
                ? 'table-actions__check-all-indeterminate'
                : 'table-actions__check-all'
            }`}
            data-testid="select-all-checkbox"
          />
          {selectedRowsCount > 0 ? (
            <div className="table-actions__selection-count">{`${SELECTED} ${selectedRowsCount}`}</div>
          ) : (
            <div className="table-actions__selection-count">
              {NO_ROWS_SELECTED}
            </div>
          )}
        </div>
      )}
      {isDownloadable && selectedRowsCount > 0 && (
        <div className="table-actions__download">
          <button
            className="table-actions__download-button"
            onClick={onDownload}
            disabled={!areDownloadRulesMeet && downloadRules.disableButton}
            data-testid="download-button"
          >
            <DonwloadIcon
              className="table-actions__download-icon"
              title="download-icon"
            />
            <span className="table-actions__download-label">
              {DOWNLOAD_LABEL}
            </span>
          </button>
          {!areDownloadRulesMeet && (
            <div className="table-actions__download-error-message" role="alert">
              {downloadRules.errorMessage}
            </div>
          )}
        </div>
      )}
    </div>
  );
};

TableActions.propTypes = {
  isSelectable: PropTypes.bool,
  isDownloadable: PropTypes.bool,
  selectedRowsCount: PropTypes.number,
  onSelectAllt: PropTypes.func,
  onDownloadt: PropTypes.func,
  areAllRowsSelected: PropTypes.bool,
  downloadRules: PropTypes.shape({
    disableButton: PropTypes.bool,
    rule: PropTypes.func,
    errorMessage: PropTypes.string,
  }),
  areDownloadRulesMeet: PropTypes.bool,
};

export default TableActions;
