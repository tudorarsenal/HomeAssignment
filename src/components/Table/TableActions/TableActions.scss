.table-actions {
  display: flex;
  padding: 20px 10px;
  font-size: 20px;
  border: 1px solid rgb(237, 237, 237);
  background-color: white;

  &__selection-count {
    text-align: center;
    margin-left: 10px;
  }

  &__select {
    flex: 1;
    display: flex;
    align-items: baseline;
    margin-right: 20px;
  }

  &__check-all-indeterminate {
    filter: sepia(100%) hue-rotate(190deg) saturate(500%);
  }

  &__download {
    flex: 5;
    display: flex;
  }

  &__download-label {
    margin-right: 10px;
  }

  &__download-button {
    display: flex;
    align-items: center;
    cursor: pointer;
    background-color: white;
    border: none;
    padding: 0;
    font-size: 20px;
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto',
      'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans',
      'Helvetica Neue', sans-serif;
    font-weight: normal;
  }

  &__download-error-message {
    align-self: center;
    font-size: 18px;
    margin-left: 20px;
  }

  &__download-icon {
    align-self: center;
    margin: 0 10px;
    width: 20px;
    height: 20px;
  }
}

@media screen and (max-width: 992px) {
  .table-actions__download-button,
  .table-actions__download-error-message {
    font-size: 14px;
  }

  .table-actions__selection-count {
    font-size: 16px;
  }
}

@media screen and (max-width: 800px) {
  .table-actions__download {
    flex: 4;
  }

  .table-actions__download-button {
    font-size: 12px;
    margin-left: 10px;
  }

  .table-actions__download-error-message {
    font-size: 12px;
  }

  .table-actions__select {
    margin-right: 10px;
  }

  .table-actions__download-icon {
    width: 15px;
    height: 15px;
    margin-right: 5px;
  }

  .table-actions__selection-count {
    font-size: 14px;
  }
}
