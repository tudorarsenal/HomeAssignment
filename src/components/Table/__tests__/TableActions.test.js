import renderer from 'react-test-renderer';

import TableActions from '../TableActions/TableActions';

describe('TableActions', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<TableActions isSelectable={true} selectedRowsCount={0} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with indeterminate state for select all', () => {
    const tree = renderer
      .create(<TableActions isSelectable={true} selectedRowsCount={2} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with select all checked', () => {
    const tree = renderer
      .create(
        <TableActions
          isSelectable={true}
          selectedRowsCount={5}
          areAllRowsSelected={true}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with download error', () => {
    const downloadRules = {
      disableButton: false,
      errorMessage: 'Requirements not meet',
    };

    const tree = renderer
      .create(
        <TableActions
          isSelectable={true}
          selectedRowsCount={2}
          isDownloadable={true}
          areDownloadRulesMeet={false}
          downloadRules={downloadRules}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
