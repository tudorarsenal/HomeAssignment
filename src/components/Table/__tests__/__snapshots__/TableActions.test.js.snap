// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableActions renders correctly 1`] = `
<div
  className="table-actions"
>
  <div
    className="table-actions__select"
  >
    <label
      className="visually-hidden"
      htmlFor="select-all"
    >
      Select all
    </label>
    <input
      className="table-actions__check-all"
      data-testid="select-all-checkbox"
      id="select-all"
      name="select-all"
      type="checkbox"
    />
    <div
      className="table-actions__selection-count"
    >
      None Selected
    </div>
  </div>
</div>
`;

exports[`TableActions renders correctly with download error 1`] = `
<div
  className="table-actions"
>
  <div
    className="table-actions__select"
  >
    <label
      className="visually-hidden"
      htmlFor="select-all"
    >
      Select all
    </label>
    <input
      className="table-actions__check-all-indeterminate"
      data-testid="select-all-checkbox"
      id="select-all"
      name="select-all"
      type="checkbox"
    />
    <div
      className="table-actions__selection-count"
    >
      Selected 2
    </div>
  </div>
  <div
    className="table-actions__download"
  >
    <button
      className="table-actions__download-button"
      data-testid="download-button"
      disabled={false}
    >
      <svg
        className="table-actions__download-icon"
        title="download-icon"
      >
        downloadIcon.svg
      </svg>
      <span
        className="table-actions__download-label"
      >
        Download Selected
      </span>
    </button>
    <div
      className="table-actions__download-error-message"
      role="alert"
    >
      Requirements not meet
    </div>
  </div>
</div>
`;

exports[`TableActions renders correctly with indeterminate state for select all 1`] = `
<div
  className="table-actions"
>
  <div
    className="table-actions__select"
  >
    <label
      className="visually-hidden"
      htmlFor="select-all"
    >
      Select all
    </label>
    <input
      className="table-actions__check-all-indeterminate"
      data-testid="select-all-checkbox"
      id="select-all"
      name="select-all"
      type="checkbox"
    />
    <div
      className="table-actions__selection-count"
    >
      Selected 2
    </div>
  </div>
</div>
`;

exports[`TableActions renders correctly with select all checked 1`] = `
<div
  className="table-actions"
>
  <div
    className="table-actions__select"
  >
    <label
      className="visually-hidden"
      htmlFor="select-all"
    >
      Select all
    </label>
    <input
      checked={true}
      className="table-actions__check-all"
      data-testid="select-all-checkbox"
      id="select-all"
      name="select-all"
      type="checkbox"
    />
    <div
      className="table-actions__selection-count"
    >
      Selected 5
    </div>
  </div>
</div>
`;
